central/authprovider/service/service_impl_test.go

```diff
@@ -0,0 +1,118 @@
+//go:build sql_integration
+
+package service
+
+import (
+	"context"
+	"testing"
+
+	authProviderDataStore "github.com/stackrox/rox/central/authprovider/datastore"
+	groupDataStore "github.com/stackrox/rox/central/group/datastore"
+	roleDataStore "github.com/stackrox/rox/central/role/datastore"
+	roleMapper "github.com/stackrox/rox/central/role/mapper"
+	userDataStore "github.com/stackrox/rox/central/user/datastore"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/auth/authproviders"
+	openshiftAuth "github.com/stackrox/rox/pkg/auth/authproviders/openshift"
+	authTokenMocks "github.com/stackrox/rox/pkg/auth/tokens/mocks"
+	"github.com/stackrox/rox/pkg/postgres/pgtest"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/mock/gomock"
+)
+
+func TestAuthProviderService(t *testing.T) {
+	suite.Run(t, new(authProviderServiceTestSuite))
+}
+
+type authProviderServiceTestSuite struct {
+	suite.Suite
+
+	db       *pgtest.TestPostgres
+	mockCtrl *gomock.Controller
+
+	service Service
+}
+
+func (s *authProviderServiceTestSuite) SetupSuite() {
+	t := s.T()
+	db := pgtest.ForT(t)
+	s.db = db
+	s.mockCtrl = gomock.NewController(t)
+
+	tokenIssuerFactory := authTokenMocks.NewMockIssuerFactory(s.mockCtrl)
+	tokenIssuerFactory.EXPECT().CreateIssuer(gomock.Any(), gomock.Any()).AnyTimes().Return(nil, nil)
+
+	roleDS, err := roleDataStore.GetTestPostgresDataStore(t, db)
+	s.Require().NoError(err)
+	authProviderDS := authProviderDataStore.GetTestPostgresDataStore(t, db)
+	groupDS := groupDataStore.GetTestPostgresDataStore(t, db, roleDS, authProviderDS)
+	userDS := userDataStore.GetTestDataStore(t)
+	mapperFactory := roleMapper.NewStoreBasedMapperFactory(groupDS, roleDS, userDS)
+	providerRegistry := authproviders.NewStoreBackedRegistry(
+		"/sso/",
+		"/auth/response/generic",
+		authProviderDS,
+		tokenIssuerFactory,
+		mapperFactory,
+	)
+
+	ctx := sac.WithAllAccess(context.Background())
+	err = providerRegistry.RegisterBackendFactory(
+		ctx,
+		openshiftAuth.TypeName,
+		openshiftAuth.NewTestFactoryCreator(t),
```

Comment:
nitpick: I would decouple these tests from specific auth provider implementation.

Maybe we can do the following:
- mock `authproviders.BackendFactory`
- instead of using `openshiftAuth.TypeName` - just use `test`
- and validate in the test that `CreateBackend` on mocked factory is called only 2 times

Reply:
To be honest, I'd keep this specific test, since it is the reproducer for the issue at the origin of the code change.

I also like the idea of abstracting the test from any auth provider implementation. I'll add a generic mocked test for that.